src/validations/html/rules/MismatchedTag.js

```diff
@@ -0,0 +1,139 @@
+import {List, Map, Record, Set} from 'immutable';
+import Code from '../rules/Code';
+
+const Location = Record({row: 0, column: 0});  // eslint-disable-line new-cap
+
+export default class {
+  constructor() {
+    this._openTagStack = List.of();
+    this._mismatchedTagErrorsByLocation = Map.of();
+    this._unopenedTagLocations = Set.of();
+    this._unopenedTagQueuesByName = Map.of();
+    this._foundSpareUnopenedTag = false;
+  }
+
+  openTag(location, {name}) {
+    this._openTagStack = this._openTagStack.push({location, name});
+  }
+
+  closeTag(location, name) {
+    const locationKey = new Location(location);
+
+    const openIndex = this._openTagStack.findLastIndex(
+      openTag => openTag.name === name);
+    const opened = openIndex >= 0;
+
+    // Assume any open tags after the matching open tag are mismatched. If the
+    // source is valid HTML, then no other open tags should exist. If no
+    // matching open tag exists, then all open tags are mismatched.
+    const mismatchedTags =
+      opened ? this._openTagStack.slice(openIndex + 1) : this._openTagStack;
+    this._mismatchedTagErrorsByLocation =
+      this._mismatchedTagErrorsByLocation.set(
+        locationKey, {name, mismatchedTags});
+
+    if (opened) {
+      this._openTagStack = this._openTagStack.slice(0, openIndex);
+    } else {
+      this._unopenedTagLocations = this._unopenedTagLocations.add(locationKey);
+      this._unopenedTagQueuesByName =
+        this._unopenedTagQueuesByName.update(
+          name, List.of(), tags => tags.push({name, location}));
+
+      // If no matching open tag exists, then assume later close tags could
+      // match some open tag instead of emptying openTagStack.
+    }
+  }
+
+  // If any spare unopened tag exists, diagnose them and only yield
+  // MISPLACED_CLOSE_TAG.
+  //
+  // If no spare exists:
+  //
+  // A mismatched close tag that has been opened means the case
+  //
+  // <div>
+  //   <span>
+  // </div>
+  //
+  // so yield UNCLOSED_TAG.
+  //
+  // A mismatched close tag that has yet to be opened means the case
+  //
+  // <div>
+  //   </span>
+  // </div>
+  //
+  // so yield UNOPENED_TAG.
+  //
+  // Note: MISPLACED_CLOSE_TAG suggestions invalidate UNCLOSED_TAG and
+  // UNOPENED_TAG suggestions by moving other mismatched close tags.
+  // Diagnosing UNCLOSED_TAG and UNOPENED_TAG accurately would require another
+  // pass after all the MISPLACED_CLOSE_TAG suggestions are resolved.
+  * done() {
+    yield* this._yieldMisplacedCloseTags();
+    if (this._foundMisplacedCloseTag) {
+      return;
+    }
+
+    // Process the remaining errors.
+    for (const [location, {name, mismatchedTags}]
+         of this._mismatchedTagErrorsByLocation) {
+      if (this._unopenedTagLocations.has(location)) {
+        yield {
+          code: Code.UNOPENED_TAG,
+          closeTag: {
+            name,
+            location: {
+              row: location.get('row'),
```

Comment:
@tenyoung795 Popcode doesn’t support _any_ version of IE (because IE doesn’t provide a way to make sandboxed client-side iframes), and even if it did, it sure as heck wouldn’t support IE8 ; )

The source of truth for “what browsers do we support” lives in [`browsers.json`](https://github.com/popcodeorg/popcode/blob/master/config/browsers.json)